cmd/soroban-rpc/internal/db/ledgerentry.go

```diff
@@ -40,6 +42,53 @@ type ledgerEntryWriter struct {
 	maxBatchSize    int
 }
 
+func (l ledgerEntryWriter) ExtendLedgerEntry(key xdr.LedgerKey, expirationLedgerSeq xdr.Uint32) error {
+	// TODO: How do we figure out the current expiration? We might need to read
+	// from the DB, but in the case of creating a new entry and immediately
+	// extending it, or extending multiple times in the same ledger, the
+	// expirationLedgerSeq might be buffered but not flushed yet.
+	if key.Type != xdr.LedgerEntryTypeContractCode && key.Type != xdr.LedgerEntryTypeContractData {
+		panic("ExtendLedgerEntry can only be used for contract code and data")
+	}
+
+	encodedKey, err := encodeLedgerKey(l.buffer, key)
+	if err != nil {
+		return err
+	}
+
+	var entry xdr.LedgerEntry
+	var existing string
+	// See if we have a pending (unflushed) update for this key
+	queued := l.keyToEntryBatch[encodedKey]
+	if queued != nil && *queued != "" {
+		existing = *queued
+	} else {
+		// Nothing in the flush buffer. Load the entry from the db
+		err = sq.StatementBuilder.RunWith(l.stmtCache).Select("entry").From(ledgerEntriesTableName).Where(sq.Eq{"key": encodedKey}).QueryRow().Scan(&existing)
+		if err == sql.ErrNoRows {
+			return fmt.Errorf("no entry for key %q in table %q", base64.StdEncoding.EncodeToString([]byte(encodedKey)), ledgerEntriesTableName)
+		} else if err != nil {
+			return err
+		}
+	}
+
+	// Unmarshal the existing entry
+	if err := xdr.SafeUnmarshal([]byte(existing), &entry); err != nil {
+		return err
+	}
+
+	// Update the expiration
```

Comment:
should we add validation here that `entry.Data.Type == key.Type` ?

Reply:
No? The key and data are different things?

Reply:
my understanding ( which could be wrong ) is that the entry contains multiple fields and that the key is one of them. and if that key is supposed to be identical to the "main" ledger key, then it would make sense to perform some sort of sanity validation.

Reply:
Ah, right. Yes, we should do that in the `Upsert` method though, so it is done pre-emptively, not just on extension. Doing that cleanly means a refactoring PR to the `go` repo, which is in [this new PR.](https://github.com/stellar/go/pull/4942)

Reply:
Actually, we can simplify the UpsertLedgerEntry method, since the key is derived from the entry, we don't need to pass in both. Then there is nothing to validate.